discovery/openstack/openstack.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package openstack
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	"github.com/gophercloud/gophercloud/pagination"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"golang.org/x/net/context"
+
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/util/strutil"
+)
+
+const (
+	openstackLabel               = model.MetaLabelPrefix + "openstack_"
+	openstackLabelInstanceID     = openstackLabel + "instance_id"
+	openstackLabelInstanceName   = openstackLabel + "instance_name"
+	openstackLabelInstanceState  = openstackLabel + "instance_state"
+	openstackLabelInstanceFlavor = openstackLabel + "instance_flavor"
+	openstackLabelTag            = openstackLabel + "tag_"
+)
+
+var (
+	openstackSDRefreshFailuresCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "prometheus_sd_openstack_refresh_failures_total",
+			Help: "The number of OpenStack-SD scrape failures.",
+		})
+	openstackSDRefreshDuration = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Name: "prometheus_sd_openstack_refresh_duration_seconds",
+			Help: "The duration of a OpenStack-SD refresh in seconds.",
+		})
+)
+
+func init() {
+	prometheus.MustRegister(openstackSDRefreshFailuresCount)
+	prometheus.MustRegister(openstackSDRefreshDuration)
+}
+
+// Discovery periodically performs OpenStack-SD requests. It implements
+// the TargetProvider interface.
+type Discovery struct {
+	os       *gophercloud.AuthOptions
```

Comment:
`os` is a bit of an unclear name, it could mean "operating system", "openstack", or "options"... I guess in this context it's meant to mean "options"? Maybe name it `authOpts`?